### ISSUE TITLE
Using `partial_eval_jaxpr` to simulate yield points

### ISSUE THREAD

**Issue by thisiscam** [google/jax#3936]
Hi all,

I've been trying to use https://github.com/google/jax/blob/master/jax/interpreters/partial_eval.py#L576 to simulate a yield point. 
The idea was to define a new primitive `yield_p`, which is just the identity function, but during partial evaluation, always treat its output as unknown regardless of the input. 
From the documentation in the link, I'm guessing that partial_eval should be able to "split" my jaxpr in half, where the first half computes everything up to the yield_p, and the second half computes everything after. 

However, below is my minimal attempt for the above idea, and reproduces a weird behavior that I wasn't able to figure out what partial_eval_jaxpr is doing.
 
```
import jax.core as core
import jax.interpreters.partial_eval as pe
from jax import source_info_util
from jax import make_jaxpr
import jax.numpy as np

yield_p = core.Primitive("yield_prim")
yield_p.def_impl(lambda x: x)
yield_p.def_abstract_eval(lambda x: x)
def do_yield(x):
  return yield_p.bind(x)

def yield_p_partial_eval_rule(trace: pe.JaxprTrace, tracer: pe.JaxprTracer,
                              **params):
  tracer = trace.instantiate_const(tracer)
  out_aval = yield_p.abstract_eval(tracer.aval, **params)
  source = source_info_util.current()
  out_tracer = pe.JaxprTracer(trace, pe.PartialVal.unknown(out_aval), None)
  out_tracer.recipe = pe.new_eqn_recipe([tracer], [out_tracer], yield_p, params,
                                        source)
  return out_tracer
pe.custom_partial_eval_rules[yield_p] = yield_p_partial_eval_rule


def f1(x):
  return do_yield(x) * 2


def f2(x):
  return do_yield(np.sin(x)) * 2


def test_yield_func(f):
  print('-----' + f.__name__ + "-----")
  jaxpr = make_jaxpr(f)(1.)
  (jaxpr_known, jaxpr_unknown,
   out_unknowns) = pe.partial_eval_jaxpr(jaxpr, [False], True, None)

  print("jaxpr:", jaxpr)
  print("jaxpr_known:", jaxpr_known)
  print("jaxpr_unknown:", jaxpr_unknown)
  print("out_unknowns:", out_unknowns)
  print("")


test_yield_func(f1)
test_yield_func(f2)
```

As above, I have overwritten `pe.custom_partial_eval_rules` for `yield_p` to always produce unknown PartialVal. 
I have defined two simple functions `f1` and `f2`, where the only difference is `f2` does some computation before passing to yield.

The weird part is in the output:
```
-----f1-----
jaxpr: { lambda  ; a.
  let b = yield_prim a
      c = mul b 2.0
  in (c,) }
jaxpr_known: { lambda  ; a.
  let 
  in (*, a) }
jaxpr_unknown: { lambda  ; a b.
  let c = yield_prim b
      d = mul c 2.0
  in (d,) }
out_unknowns: [True]

-----f2-----
jaxpr: { lambda  ; a.
  let b = sin a
      c = yield_prim b
      d = mul c 2.0
  in (d,) }
jaxpr_known: { lambda  ; a.
  let b = sin a
      c = yield_prim b
      d = mul c 2.0
  in (*, d) }
jaxpr_unknown: { lambda  ; a b.
  let 
  in (b,) }
out_unknowns: [True]
```

It turns out that `jaxpr_known` is empty for `f1`, which is what I expected. But `jaxpr_known` is the full jaxpr of `f2`, which got me very confused. 

I'm thinking there're three ways to explain this behavior:
1) `partial_eval_jaxpr` doesn't really look at the output of custom partial eval rules to determine which part of jaxpr is known vs unknown; instead it only looks at dependencies of the unknowns of the input. This appears to be what the docstring says -- which is fine, but quite different from my understanding of partial evaluation (*which could be wrong). However, this is odd since for `f1` it decided to treat the entire jaxpr as unknown, and seems to have not ignored my custom rule.
2) `partial_eval_jaxpr` was intended to look at the outputs of custom partial eval, but something might be buggy internally with JAX.
3) Something is wrong with my implementation.

Currently, seems like 3) is most likely what happened; so I'd be very appreciated for a JAX expert to look at why this happens. Or maybe teach me what I missed? Many thanks in advance!

**Comment by thisiscam** [comment]
To add, this seems somewhat related to #1660, which provides `yield_p` similar to what I desire. Maybe the fact that it is much more complicated than using `partial_eval` implies that `partial_eval` is not intended to be abused this way :)

**Comment by apaszke** [comment]
This is a very interesting case, but it's also something that the current implementation is not really prepared to handle, because it is based entirely on data dependencies. To construct `jaxpr_unknown` it will start executing your program with the sole input marked as known. Because the output of `jnp.sin` is also known, it does not get emitted into the trace and the system _does not bother marking the result as participating in the partial eval_, because it is assumed that a known value can never give rise to an unknown value. This last bit is important, because form that point on, the result of `jnp.sin` is indistinguishable from any value your function might have closed over. In particular, the input to `yield_p` is not marked as participating in partial_eval, so `yield_p` can be even ran using its impl rule.

To sum up, I don't think this is a bug, beacuse the partial_eval used to be designed with data dependencies in mind, and in this world you can never have an unknown value pop up out of nowhere. But we could try to change it (especially with the recent omnistaging changes) such that it would be able to handle the use-case you've suggested. WDYT @mattjj?

**Comment by thisiscam** [comment]
@apaszke 
Thanks for the explanation! That clarifies a lot.

Let me echo  back my guess on why for `f1` the entire jaxpr is unknown: the `yield_p` is not marked as participating in the partial eval, therefore it's output value, which is an unknown, just propagates and partial eval decided that everything is unknown?  

Anyways, after looking at #1660 I'm more curious about the progress on tagging as that should properly implement a yield point, which provides a cleaner API than what I was trying to do. But I do think that approach have a problem similar in vein, since implementations in both #1660 and #1803 both require "plumbing" a scope variable, which carries around the top level dependency to make sure subexpressions are staged properly.

**Comment by apaszke** [comment]
Yeah `f1` worked only because `yield_p` is applied to the function argument. The arguments are always marked as participating in partial eval, but when you run an op on known arguments the outputs will not be marked as participating. I hacked together a prototype (really only a few lines changed) that would avoid dropping this info and it does produce the right output for me now. Unfortunately it seems to be failing some tests, so it might not be that straightforward...

**Comment by jekbradbury** [comment]
cc @sharadmv; this user might want to use Oryx!

**Comment by thisiscam** [comment]
Hi @jekbradbury thanks for pointing out that library! It indeed looks very interesting and I may potentially make use of it.

Since I see that Oryx provides `reap`/`plant` functions which are equivalent to `collect`/`inject` transforms you gave in #1660, is the plan that these functions will be moved to Oryx permanently? 

Also, it seems like `reap`/`plant` ignores tagged value's intermediate gradient information, and back in #1660 you mentioned it was a design choice that was not clear what is the best. Were you able to progress onto a newer design, and might share some new insight? Thanks!

**Comment by apaszke** [comment]
This was an interesting discussion, but I'll close the issue for the sake of keeping our repo tidy. Feel free to post on our discussions forum if you have any further questions!
